File: static/solutions/java/src/main/java/com/epi/UniformRandomNumberGeneration.java
package com.epi;

import java.util.Random;

public class UniformRandomNumberGeneration {
  static int zeroOneRandom() {
    Random gen = new Random();
    return gen.nextInt(2);
  }

  // @include
  public static int uniformRandom(int lowerBound, int upperBound) {
    int numberOfOutcomes = upperBound - lowerBound + 1, result;
    do {
      result = 0;
      for (int i = 0; (1 << i) < numberOfOutcomes; ++i) {
        // zeroOneRandom() is the provided random number generator.
        result = (result << 1) | zeroOneRandom();
      }
    } while (result >= numberOfOutcomes);
    return result + lowerBound;
  }
  // @exclude

  public static void main(String[] args) {
    Random gen = new Random();
    for (int times = 0; times < 1000; ++times) {
      int a, b;
      if (args.length == 2) {
        a = Integer.parseInt(args[0]);
        b = Integer.parseInt(args[1]);
      } else {
        a = gen.nextInt(100);
        b = gen.nextInt(100) + a + 1;
      }
      int x = uniformRandom(a, b);
      System.out.println(String.format("a = %d, b = %d", a, b));
      System.out.println(String.format("random result = %d", x));
      assert x >= a&& x <= b;
    }
  }
}

File: static/cpp/Uniform_random_number_generation.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <cmath>
#include <iostream>
#include <random>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::random_device;
using std::uniform_int_distribution;

int ZeroOneRandom() {
  default_random_engine gen((random_device())());
  uniform_int_distribution<int> dis(0, 1);
  return dis(gen);
}

// @include
int UniformRandom(int lower_bound, int upper_bound) {
  int number_of_outcomes = upper_bound - lower_bound + 1, result;
  do {
    result = 0;
    for (int i = 0; (1 << i) < number_of_outcomes; ++i) {
      // ZeroOneRandom() is the provided random number generator.
      result = (result << 1) | ZeroOneRandom();
    }
  } while (result >= number_of_outcomes);
  return result + lower_bound;
}
// @exclude

int main(int argc, char* argv[]) {
  default_random_engine gen((random_device())());
  for (int times = 0; times < 1000; ++times) {
    int a, b;
    if (argc == 3) {
      a = atoi(argv[1]), b = atoi(argv[2]);
    } else {
      uniform_int_distribution<int> a_dis(0, 99);
      a = a_dis(gen);
      uniform_int_distribution<int> b_dis(a + 1, a + 100);
      b = b_dis(gen);
    }
    int x = UniformRandom(a, b);
    cout << "a = " << a << " b = " << b << endl;
    cout << "random result = " << x << endl;
    assert(x >= a && x <= b);
  }
  return 0;
}
